srcts/src/shiny/bind.ts

```diff
@@ -120,14 +140,34 @@ const bindingsRegistry = (() => {
       })
       .join("\n");
 
+    let txtVerb = "Duplicate";
+    let txtNoun = "input/output";
+    if (problems.has("input") && problems.has("output")) {
+      // base case
+    } else if (problems.has("input")) {
+      txtNoun = "input";
+    } else if (problems.has("output")) {
+      txtNoun = "output";
+    } else if (problems.has("shared")) {
+      txtVerb = "Shared";
+    }
+
+    const txtIdsWere = duplicateIds.size == 1 ? "ID was" : "IDs were";
+    const headline = `${txtVerb} ${txtNoun} ${txtIdsWere} found`;
+    const message = `The following ${txtIdsWere} used for more than one ${
+      problems.has("shared") ? "input/output" : txtNoun
+    }:\n${duplicateIdMsg}`;
+
+    if (status === "warning") {
+      return {
+        status: "warning",
+        event: new ShinyClientMessageEvent({ headline, message }),
+      };
+    }
+
     return {
       status: "error",
-      error: new ShinyClientError({
-        headline: "Duplicate input/output IDs found",
-        message: `The following ${
-          duplicateIds.size === 1 ? "ID was" : "IDs were"
-        } repeated:\n${duplicateIdMsg}`,
-      }),
+      error: new ShinyClientError({ headline, message }),
     };
```

Comment:
Seems `status: "error"` is no longer possible/relevant for `checkValidity`? In that case, I'd be favor of removing it completely, or at least a healthy comment about how we imagine it being relevant in the future

Reply:
Good point, I had left this because I didn't want to change the scope of what happens with `checkValidity()` too much. But I've refactored it now for clarity, which helps a lot and brings all of the logic into `checkValidity()`.